scripts/wildcard-domain-processor/wildcard-expander.ts

```diff
@@ -0,0 +1,263 @@
+/* eslint-disable no-restricted-syntax,no-await-in-loop */
+import path from 'path';
+
+import agtree, {
+    AnyRule,
+    CosmeticRule, DomainListParser,
+    NetworkRule,
+    RuleParser,
+} from '@adguard/agtree';
+
+import { findFilterFiles, readFile, writeFile } from './file-utils';
+import { WildcardDomains } from './wildcard-domains-updater';
+import { DOMAIN_MODIFIERS } from './domain-extractor';
+import { utils } from './utils';
+import { updateContentChecksum } from '../checksum';
+
+/**
+ * Expands wildcards in a network rule AST.
+ * @param ast - The network rule AST to process.
+ * @param wildcardDomains - A map of wildcard domains to their non-wildcard equivalents.
+ * @returns The updated network rule AST with expanded wildcards, or null if no valid domains are left.
+ */
+function expandWildcardsInNetworkRules(
+    ast: NetworkRule,
+    wildcardDomains: WildcardDomains,
+): NetworkRule | null {
+    if (!ast.modifiers) {
+        return ast;
+    }
+
+    const modifiers = ast.modifiers.children;
+    const newPermittedDomains = new Map();
+    const newRestrictedDomains = new Map();
+    let hadWildcard = false;
+    const newModifiers = [];
+
+    for (const modifier of modifiers) {
+        if (!DOMAIN_MODIFIERS.includes(modifier.modifier.value) || !modifier.value) {
+            newModifiers.push(modifier);
+            continue;
+        }
+
+        const domainList = DomainListParser.parse(modifier.value.value, agtree.PIPE_MODIFIER_SEPARATOR);
+
+        for (const domain of domainList.children) {
+            if (utils.isWildcardDomain(domain.value)) {
+                hadWildcard = true;
+                const nonWildcardDomains = wildcardDomains[domain.value];
```

Comment:
I guess it is possible that `nonWildcardDomains` is undefined, isn't it?
e.g. when `wildcard_domains.json` hasn't been updated for some time and some new wildcard domain rule has been added

should error be thrown in this case?

Reply:
Nice catch.
Actually, we can throw an error here, so it will force the update of the wildcard_domains JSON.
However, in my opinion, this should be done more gracefully since the script that updates wildcard domains can run for 2-3 hours.

Reply:
We should ignore those and keep wildcard domains in the filter list in this case.

This is a normal situation and it will occur.